### PR TITLE
[Migrate]barcodes: Migrate to new api.

### PR DESCRIPTION
**Pull request by rpi-odoo**
Task:
https://www.odoo.com/web#id=13727&view_type=form&model=project.task&menu_id=3942&action=327
Pad:
https://pad.odoo.com/p/openerp-project.task-DHQH4CJTRW

### CODE REVIEW

File: addons/barcodes/models/barcode_nomenclature.py [hunk 1]
```diff
@@ -1,34 +1,29 @@
...
+
 import logging
 import re
 
-import openerp
-from openerp import tools, models, fields, api
-from openerp.osv import fields, osv
-from openerp.tools.translate import _
-from openerp.exceptions import ValidationError
+from openerp import models, fields
```
Comment (WHPODOO) [inline addons/barcodes/models/barcode_nomenclature.py]: Import should be in alphabetic order like : 
 ```
from openerp import fields, models
```

File: addons/barcodes/models/barcode_rule.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import logging
+import re
+
+from openerp import models, fields, api
```
Comment (WHPODOO) [inline addons/barcodes/models/barcode_rule.py]: Import should be in alphabetic order like : 
 ```
from openerp import api, fields, models
```

File: addons/barcodes/models/barcode_rule.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+from openerp.tools.translate import _
```
Comment (WHPODOO) [inline addons/barcodes/models/barcode_rule.py]: No need to import separate melt with above import like : 
```
from openerp import api, fields, models, _
```

File: addons/barcodes/models/barcode_nomenclature.py [hunk 4]
```diff
@@ -1,34 +1,29 @@
...
-        'name': fields.char('Nomenclature Name', size=32, required=True, help='An internal identification of the barcode nomenclature'),
-        'rule_ids': fields.one2many('barcode.rule','barcode_nomenclature_id','Rules', help='The list of barcode rules'),
-        'upc_ean_conv': fields.selection(UPC_EAN_CONVERSIONS, 'UPC/EAN Conversion', required=True,
-            help='UPC Codes can be converted to EAN by prefixing them with a zero. This setting determines if a UPC/EAN barcode should be automatically converted in one way or another when trying to match a rule with the other encoding.'),
-    }
 
-    _defaults = {
-        'upc_ean_conv': 'always',
-    }
+    name = fields.Char('Nomenclature Name', required=True, help='An internal identification of the barcode nomenclature')
```
Comment (mdi-odoo) [inline addons/barcodes/models/barcode_nomenclature.py]: Why you removed the 'size' attribute  ?

File: addons/barcodes/models/barcode_rule.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+from odoo.exceptions import ValidationError
+
+_logger = logging.getLogger(__name__)
+
+
+class BarcodeRule(models.Model):
+    _name = 'barcode.rule'
+    _order = 'sequence asc'
+
+    name = fields.Char('Rule Name', required=True, help='An internal identification for this barcode nomenclature rule')
```
Comment (mdi-odoo) [inline addons/barcodes/models/barcode_rule.py]: Same here .

File: addons/barcodes/models/barcode_nomenclature.py [hunk 6]
```diff
@@ -1,34 +1,29 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
 import logging
```
Comment (mdi-odoo) [inline addons/barcodes/models/barcode_nomenclature.py]: Unused import statement: logging .

Comment (rpi-odoo) [inline addons/barcodes/models/barcode_nomenclature.py]: _logger = logging.getLogger(__name__)
logging is used here.

File: addons/barcodes/models/barcode_rule.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    @api.model
+    def _get_type_selection(self):
+        return [('alias', 'Alias'), ('product', 'Unit Product')]
+
+    @api.multi
+    @api.constrains('pattern')
+    def _check_pattern(self):
+        for record in self:
+            p = record.pattern.replace("\\\\", "X").replace("\{", "X").replace("\}", "X")
```
Comment (WHPODOO) [inline addons/barcodes/models/barcode_rule.py]: Give proper variable name. What is ```p```  :question:

File: addons/barcodes/models/barcode_rule.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+class BarcodeRule(models.Model):
+    _name = 'barcode.rule'
+    _order = 'sequence asc'
+
+    name = fields.Char('Rule Name', required=True, help='An internal identification for this barcode nomenclature rule')
+    barcode_nomenclature_id = fields.Many2one('barcode.nomenclature', 'Barcode Nomenclature')
+    sequence = fields.Integer('Sequence', help='Used to order rules such that rules with a smaller sequence match first')
+    encoding = fields.Selection('_encoding_selection_list', 'Encoding', required=True, default='any', help='This rule will apply only if the barcode is encoded with the specified encoding')
+    rule_type = fields.Selection('_get_type_selection', 'Type', required=True, default='product', oldname='type')
```
Comment (hmo-odoo) [inline addons/barcodes/models/barcode_rule.py]: can you make separate commit regarding changes of field name like type ==> rule_type

File: addons/barcodes/models/barcode_rule.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+
+class BarcodeRule(models.Model):
+    _name = 'barcode.rule'
+    _order = 'sequence asc'
+
+    name = fields.Char('Rule Name', required=True, help='An internal identification for this barcode nomenclature rule')
+    barcode_nomenclature_id = fields.Many2one('barcode.nomenclature', 'Barcode Nomenclature')
+    sequence = fields.Integer('Sequence', help='Used to order rules such that rules with a smaller sequence match first')
+    encoding = fields.Selection('_encoding_selection_list', 'Encoding', required=True, default='any', help='This rule will apply only if the barcode is encoded with the specified encoding')
```
Comment (hmo-odoo) [inline addons/barcodes/models/barcode_rule.py]: why pass _encoding_selection_list method in string argument ?

File: addons/barcodes/models/barcode_rule.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    rule_type = fields.Selection('_get_type_selection', 'Type', required=True, default='product', oldname='type')
+    pattern = fields.Char('Barcode Pattern', help="The barcode matching pattern", required=True, default='.*')
+    alias = fields.Char('Alias', default='0', help='The matched pattern will alias to this barcode', required=True)
+
+    @api.model
```
Comment (hmo-odoo) [inline addons/barcodes/models/barcode_rule.py]: do not decorate

File: addons/barcodes/models/barcode_rule.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    def _encoding_selection_list(self):
+        return [
+            ('any', 'Any'),
+            ('ean13', 'EAN-13'),
+            ('ean8', 'EAN-8'),
+            ('upca', 'UPC-A'),
+        ]
+
+    @api.model
+    def _get_type_selection(self):
```
Comment (hmo-odoo) [inline addons/barcodes/models/barcode_rule.py]: do not decorate

File: addons/barcodes/models/barcode_rule.py (hunk 7 above)
Comment (hmo-odoo) [inline addons/barcodes/models/barcode_rule.py]: need to find better pattern for barcode

File: addons/barcodes/models/barcode_rule.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
+# -*- coding: utf-8 -*-
+import re
+
+from odoo import api, fields, models, _
+
+
+
+class BarcodeRule(models.Model):
```
Comment (WHPODOO) [inline addons/barcodes/models/barcode_rule.py]: what is the need of three new line :question: You follow ***PEP8*** standard then as per standard need two new line :smiley: 
use your brain

File: addons/barcodes/models/barcode_nomenclature.py [hunk 13]
```diff
@@ -141,7 +130,7 @@ def parse_barcode(self, barcode):
 
         rules = []
         for rule in self.rule_ids:
-            rules.append({'type': rule.type, 'encoding': rule.encoding, 'sequence': rule.sequence, 'pattern': rule.pattern, 'alias': rule.alias})
+            rules.append({'rule_type': rule.type, 'encoding': rule.encoding, 'sequence': rule.sequence, 'pattern': rule.pattern, 'alias': rule.alias})
```
Comment (msh-odoo) [inline addons/barcodes/models/barcode_nomenclature.py]: While testing barcode interface(Inside stock.picking form), scanning barcode throws traceback for type, note that type is changed to rule_type

Comment (rpi-odoo) [inline addons/barcodes/models/barcode_nomenclature.py]: @msh-odoo Ok.. :+1:

File: addons/barcodes/models/barcode_nomenclature.py [hunk 14]
```diff
@@ -155,12 +144,12 @@ def parse_barcode(self, barcode):
...
             match = self.match_pattern(cur_barcode, rule['pattern'])
             if match['match']:
-                if rule['type'] == 'alias':
+                if rule['rule_type'] == 'alias':
                     barcode = rule['alias']
                     parsed_result['code'] = barcode
                 else:
                     parsed_result['encoding'] = rule['encoding']
-                    parsed_result['type'] = rule['type']
+                    parsed_result['rule_type'] = rule['rule_type']
```
Comment (msh-odoo) [inline addons/barcodes/models/barcode_nomenclature.py]: Now your parsed result sending key as rule_type, so you also need to do changes from where this parse_barcode is called otherwise it may throw traceback while accessing type from parsed result

File: addons/barcodes/models/barcode_nomenclature.py [hunk 15]
```diff
@@ -141,7 +130,7 @@ def parse_barcode(self, barcode):
 
         rules = []
         for rule in self.rule_ids:
-            rules.append({'type': rule.type, 'encoding': rule.encoding, 'sequence': rule.sequence, 'pattern': rule.pattern, 'alias': rule.alias})
+            rules.append({'rule_type': rule.type, 'encoding': rule.encoding, 'sequence': rule.sequence, 'pattern': rule.pattern, 'alias': rule.alias})
 
         for rule in rules:
```
Comment (msh-odoo) [inline addons/barcodes/models/barcode_nomenclature.py]: Why not iterating over self.rule_ids and remove line 131 to 133

File: addons/barcodes/models/barcode_rule.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    def _encoding_selection_list(self):
+        return [
+                ('any', 'Any'),
+                ('ean13', 'EAN-13'),
+                ('ean8', 'EAN-8'),
+                ('upca', 'UPC-A'),
+        ]
+
+    @api.model
+    def _get_type_selection(self):
```
Comment (msh-odoo) [inline addons/barcodes/models/barcode_rule.py]: remove decoration

File: addons/barcodes/models/barcode_rule.py [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+    name = fields.Char(string='Rule Name', required=True, help='An internal identification for this barcode nomenclature rule')
+    barcode_nomenclature_id = fields.Many2one('barcode.nomenclature', string='Barcode Nomenclature')
+    sequence = fields.Integer(string='Sequence', help='Used to order rules such that rules with a smaller sequence match first')
+    encoding = fields.Selection(_encoding_selection_list, string='Encoding', required=True, default='any', help='This rule will apply only if the barcode is encoded with the specified encoding')
+    rule_type = fields.Selection(_type_selection, string='Type', required=True, default='product', oldname='type')
+    pattern = fields.Char(string='Barcode Pattern', help="The barcode matching pattern", required=True, default='.*')
+    alias = fields.Char(string='Alias', default='0', help='The matched pattern will alias to this barcode', required=True)
+
+    @api.multi
```
Comment (msh-odoo) [inline addons/barcodes/models/barcode_rule.py]: constrains should be enough here

File: addons/barcodes/models/barcode_rule.py [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    @api.constrains('pattern')
+    def _check_pattern(self):
+        p = re.sub(r'\\\\|\\{|\\}', "X", self.pattern)
```
Comment (msh-odoo) [inline addons/barcodes/models/barcode_rule.py]: Iterate over loop and check for each record

File: addons/barcodes/models/barcode_rule.py [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+
+
+class BarcodeRule(models.Model):
+    _name = 'barcode.rule'
+    _order = 'sequence asc'
+
+    def _encoding_selection_list(self):
```
Comment (msh-odoo) [inline addons/barcodes/models/barcode_rule.py]: If it is going to be static list then we can remove the function, if someone wants to add new key in selection then he can add through selection_add

File: addons/barcodes/views/barcodes_view.xml [hunk 20]
```diff
@@ -1,8 +1,7 @@
 <?xml version="1.0"?>
-<openerp>
-    <data>
+<odoo>
         <!--     BARCODE NOMENCLATURES     -->
-        <record model="ir.ui.view" id="view_barcode_nomenclature_form">
+        <record id="view_barcode_nomenclature_form" model="ir.ui.view">
```
Comment (WHPODOO) [inline addons/barcodes/views/barcodes_view.xml]: Unusual Diff :laughing:

File: addons/barcodes/views/barcodes_view.xml [hunk 21]
```diff
@@ -41,7 +40,7 @@
             </field>
         </record>
 
-        <record model="ir.ui.view" id="view_barcode_nomenclature_tree">
+        <record id="view_barcode_nomenclature_tree" model="ir.ui.view">
```
Comment (WHPODOO) [inline addons/barcodes/views/barcodes_view.xml]: Unusual Diff :laughing:

File: addons/barcodes/views/barcodes_view.xml [hunk 22]
```diff
@@ -51,7 +50,7 @@
             </field>
         </record>
 
-        <record model="ir.actions.act_window" id="action_barcode_nomenclature_form">
+        <record id="action_barcode_nomenclature_form" model="ir.actions.act_window">
```
Comment (WHPODOO) [inline addons/barcodes/views/barcodes_view.xml]: Unusual Diff :laughing:

File: addons/barcodes/views/barcodes_view.xml [hunk 23]
```diff
@@ -66,7 +65,7 @@
             </field>
         </record>
 
-        <record model="ir.ui.view" id="view_barcode_rule_form">
+        <record id="view_barcode_rule_form" model="ir.ui.view">
```
Comment (WHPODOO) [inline addons/barcodes/views/barcodes_view.xml]: Unusual Diff :laughing:

File: addons/barcodes/models/barcode_events_mixin.py [hunk 24]
```diff
@@ -11,7 +11,7 @@ class BarcodeEventsMixin(models.AbstractModel):
 
     _name = 'barcodes.barcode_events_mixin'
 
-    _barcode_scanned = fields.Char("Barcode Scanned", help="Value of the last barcode scanned.", store=False)
+    _barcode_scanned = fields.Char(string="Barcode Scanned", help="Value of the last barcode scanned.", store=False)
```
Comment (WHPODOO) [inline addons/barcodes/models/barcode_events_mixin.py]: Hey man , you are ??? please check your sense, No need to increase diff , without string it work Waw ammmmzzzzed.
I think you when you code >>> :laughing: 
codelike

File: addons/barcodes/models/barcode_nomenclature.py [hunk 25]
```diff
@@ -17,18 +11,13 @@
...
-    _defaults = {
-        'upc_ean_conv': 'always',
-    }
+    name = fields.Char(string='Nomenclature Name', required=True, help='An internal identification of the barcode nomenclature')
+    rule_ids = fields.One2many('barcode.rule', 'barcode_nomenclature_id', string='Rules', help='The list of barcode rules')
+    upc_ean_conv = fields.Selection(UPC_EAN_CONVERSIONS, string='UPC/EAN Conversion', required=True, default='always',
+        help="UPC Codes can be converted to EAN by prefixing them with a zero. This setting determines if a UPC/EAN barcode should be automatically converted in one way or another when trying to match a rule with the other encoding.")
 
     # returns the checksum of the ean13, or -1 if the ean has not the correct length, ean must be a string
     def ean_checksum(self, ean):
```
Comment (WHPODOO) [inline addons/barcodes/models/barcode_nomenclature.py]: Hey man migration means not to remove
```
cr
uid
ids
context
```
Odoo have some standards, please stick with it.
In migration need some technical improvements you have to do it.
In below method
```
def ean_checksum(self, ean):
def ean8_checksum(self,ean):
```
First variable ```code```  is use less so please remove it and directly do it like : 
 ```
if len(list(ean)) != 13:
    return -1
```
beaprogrammer

File: addons/barcodes/models/barcode_nomenclature.py [hunk 26]
```diff
@@ -139,87 +128,32 @@ def parse_barcode(self, barcode):
...
-                    parsed_result['type'] = rule['type']
+                    parsed_result['encoding'] = rule.encoding
+                    parsed_result['type'] = rule.type
                     parsed_result['value'] = match['value']
                     parsed_result['code'] = cur_barcode
-                    if rule['encoding'] == "ean13":
+                    if rule.encoding == "ean13":
                         parsed_result['base_code'] = self.sanitize_ean(match['base_code'])
-                    elif rule['encoding'] == "upca":
+                    elif rule.encoding == "upca":
```
Comment (WHPODOO) [inline addons/barcodes/models/barcode_nomenclature.py]: If you don't mind please don't make it so many operation if that same possible in a single operation.
Learn which operation occupies how much time then put it..
Its totally blind migration.
```
parsed_result = dict(
parsed_result,
encoding=rule.encoding,
type=rule.type,
value=match['value']
code=cur_barcode
)
```
:turtle: 
sorry you are doing like this.
you-can-do-your-job-your-eyes-closed

File: addons/barcodes/models/barcode_rule.py [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+
+
+class BarcodeRule(models.Model):
+    _name = 'barcode.rule'
+    _order = 'sequence asc'
+
+
+    def _get_type_selection(self):
```
Comment (WHPODOO) [inline addons/barcodes/models/barcode_rule.py]: yomma miyaaaaaa......
I think my morning is really very cool..... :laughing: 
Hey bro. @rpi-odoo which standers you are use for this coding ???
If i am wrong then correct me but tell me the name of the standards which supports after start class, define global variable then need to take 2 blank line....... ???
anigif_enhanced-28576-1412437486-1
@morning found your clone :laughing:

File: addons/barcodes/models/barcode_rule.py [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+    name = fields.Char(string='Rule Name', required=True, help='An internal identification for this barcode nomenclature rule')
+    barcode_nomenclature_id = fields.Many2one('barcode.nomenclature', string='Barcode Nomenclature')
+    sequence = fields.Integer(string='Sequence', help='Used to order rules such that rules with a smaller sequence match first')
+    encoding = fields.Selection([
+        ('any', 'Any'),
+        ('ean13', 'EAN-13'),
+        ('ean8', 'EAN-8'),
+        ('upca', 'UPC-A'),
+    ], string='Encoding', required=True, default='any', help='This rule will apply only if the barcode is encoded with the specified encoding')
+    type = fields.Selection(_type_selection, string='Type', required=True, default='product')
```
Comment (WHPODOO) [inline addons/barcodes/models/barcode_rule.py]: Its my suggestion but python have keyword that's use for reservation :laughing:  hope so you understand my comment , so if really need to change please change it. :space_invader:

File: addons/barcodes/models/barcode_rule.py [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+            findall = re.findall("[{]|[}]", p) # p does not contain escaped { or }
+            if len(findall) == 2:
+                if not re.search("[{][N]*[D]*[}]", p):
+                    raise ValidationError(_("There is a syntax error in the barcode pattern ") + record.pattern + _(": braces can only contain N's followed by D's."))
+                elif re.search("[{][}]", p):
+                    raise ValidationError(_("There is a syntax error in the barcode pattern ") + record.pattern + _(": empty braces."))
+            elif len(findall) != 0:
+                raise ValidationError(_("There is a syntax error in the barcode pattern ") + record.pattern + _(": a rule can only contain one pair of braces."))
+            elif p == '*':
+                raise ValidationError(_(" '*' is not a valid Regex Barcode Pattern. Did you mean '.*' ?"))
```
Comment (WHPODOO) [inline addons/barcodes/models/barcode_rule.py]: You don't think this method needs to optimize ???


After optimization always code --- "ohh really i am looking like this ??" :laughing:  
take it own coffee.

File: addons/barcodes/views/barcodes_templates.xml [hunk 30]
```diff
@@ -13,5 +12,4 @@
             </xpath>
```
Comment (WHPODOO) [inline addons/barcodes/views/barcodes_templates.xml]: You don't think so you have to remove closer tag ???
```
<script></script> --to-- <script />
```

File: addons/barcodes/__openerp__.py [hunk 31]
```diff
@@ -27,10 +29,9 @@
     'depends': ['web'],
     'data': [
         'data/barcodes_data.xml',
-        'barcodes_view.xml',
+        'views/barcodes_view.xml',
         'security/ir.model.access.csv',
-        'views/templates.xml',
+        'views/barcodes_templates.xml',
```
Comment (WHPODOO) [inline addons/barcodes/__openerp__.py]: Here you can arrange alphabetically.. like
```
A for Apple,
B for Bird,
---
``` 
alphabet-coloring-pages-printable-apple-and-bird

File: addons/barcodes/models/barcode_events_mixin.py [hunk 32]
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
 
-from openerp import models, fields, api
+from odoo import api, fields, models
 
 class BarcodeEventsMixin(models.AbstractModel):
```
Comment (WHPODOO) [inline addons/barcodes/models/barcode_events_mixin.py]: I think you feel to follow PEP8 standard and based on this standard need to take 2 lines between import's section and class.
[GUIDELINE](https://www.python.org/dev/peps/pep-0008/)

File: addons/barcodes/models/barcodes.py [hunk 33]
```diff
@@ -186,26 +178,18 @@ def _encoding_selection_list(self):
...
-        'pattern': '.*',
-        'encoding': 'any',
-        'alias': "0",
-    }
+    _type_selection = lambda self: self._get_type_selection()
+
+    name = fields.Char(string='Rule Name', size=32, required=True, help='An internal identification for this barcode nomenclature rule')
+    barcode_nomenclature_id = fields.Many2one('barcode.nomenclature', string='Barcode Nomenclature')
+    sequence = fields.Integer(string='Sequence', help='Used to order rules such that rules with a smaller sequence match first')
+    encoding = fields.Selection(_encoding_selection_list, string='Encoding', required=True, default='any', help='This rule will apply only if the barcode is encoded with the specified encoding')
```
Comment (xmo-odoo) [inline addons/barcodes/models/barcodes.py]: this change is incorrect, the method should still be passed as a name otherwise it can't be overridden. However the `@api.multi` on the method can be removed.

File: addons/barcodes/models/barcodes.py [hunk 34]
```diff
@@ -186,26 +178,18 @@ def _encoding_selection_list(self):
...
+    type = fields.Selection(_type_selection, string='Type', required=True, default='product')
```
Comment (xmo-odoo) [inline addons/barcodes/models/barcodes.py]: This change and the indirection through a local function is unnecessary, the method can be passed by name

Comment (xmo-odoo) [inline addons/barcodes/models/barcodes.py]: In fact the whole thing seems completely unnecessary since the new API allows adding selection options.
